Support for hexacopter motors

conquinn99 opened the pull request:
## Main Changes

- Definitions of six motor pins (using preprocessor directives)


## Tests

TBD

jamie-54 committed:
working on refactoring

jamie-54 committed:
still working on refactoring

About:
- probably we can fly (reading data from RC)
- sending data from RC to ESP (separate classes)
- working towards AH Kalman filter

jamie-54 committed:
first implementation of ah KF

alphaville committed:
json file with constant parameters
and a few more changes

jamie-54 committed:
fixing attitude control gains
set idle motor speed while flying

jamie-54 committed:
EspBridge: method to read from ESP

jamie-54 committed:
RadioReceiver: remove method that shouldnt be there

jamie-54 committed:
getting rid of legacy code

jamie-54 committed:
maybe now?

jamie-54 committed:
remove more codez

jamie-54 committed:
wip: altitude estimator and controller

jamie-54 committed:
wip: altitude hold control
no safety measures taken

jamie-54 committed:
data logging to black box works

jamie-54 committed:
update idle speed

alphaville committed:
tidying up a few thingies in main.py
add min_altitude_hold_altitude in constants.json

Commit:
start

Commit:
minor

Commit:
implementation of new kalman filter

Commit:
minor change to C matrix

Commit:
removed unnecessary file

Commit:
added cov for gnss and barom

Commit:
minor change to cov

Commit:
added back missing files

Commit:
removed unwanted files from commit

Commit:
added back deleted files

jamie-54 committed:
Merge branch 'fix/refactoring-main' of github.com:QUB-ASL/bzzz into fix/refactoring-main

jamie-54 committed:
recent flight data and data ploter

jamie-54 committed:
saving data for now to be deleted at a later date

jamie-54 committed:
saving data for now to be deleted at a later date

jamie-54 committed:
remove unnecessary files

jamie-54 committed:
remove more unnecessary files

jamie-54 committed:
remove unused file and print statments

jamie-54 committed:
Fix radio data so main script doesn't stop when reciver is disconected

jamie-54 committed:
add angles file

jamie-54 committed:
correct variable names from cm to meters

jamie-54 committed:
add documentation

jamie-54 committed:
increase increase refrence

jamie-54 committed:
recent data

jamie-54 committed:
add sensors to main.py

jamie-54 committed:
Merge branch 'feature/202-Initialisation_GPS_barometer_altitude' into fix/refactoring-main

jamie-54 committed:
remove unnecessary files

jamie-54 committed:
fix small errors to kalman filter

jamie-54 committed:
Merge branch 'fix/refactoring-main' of github.com:QUB-ASL/bzzz into fix/refactoring-main

jamie-54 committed:
Merge branch 'dev' into fix/new_kalman_filter

jamie-54 committed:
Merge branch 'dev' into fix/refactoring-main

jamie-54 committed:
Merge branch 'dev' into fix/refactoring-main

jamie-54 committed:
Merge branch 'fix/refactoring-main' into fix/new_kalman_filter

jamie-54 committed:
update Kalman filter for varrying values of R matrix  etc to match paper

jamie-54 committed:
Merge pull request #228 from QUB-ASL/fix/new_kalman_filter

new kalman filter on quadcopter for altitude hold

jamie-54 committed:
Merge branch 'dev' into fix/refactoring-main

jamie-54 committed:
fix issues with saving files

jamie-54 committed:
minor changes to obey pep8 rules

conquinn99 committed:
motor pins updated

conquinn99 commented on include/config.hpp:
```diff
@@ -39,10 +43,19 @@
 /*
  * Motors config
  */
-#define FRONT_LEFT_ESC_PIN 25
```
This is a comment

conquinn99 committed:
Added code to enable two extra motors (middle left and middle right) for use in hexacopter using #ifdef function.

alphaville commented on src/main.cpp:
```diff
@@ -22,38 +22,25 @@ float yawReferenceRad = 0.0;
-bool wasKill=0;
-bool isKill=0;
+
+#if UAV_TYPE == UAV_TYPE_HEXACOPTER
```
I suppose "ML" and "MR" stands for "mid left" and "mid right". Put a comment.

alphaville commented on src/main.cpp:
```diff
@@ -22,38 +22,25 @@ float yawReferenceRad = 0.0;
 unsigned long timestampLastKill = 0;
 bool isThrottleStickDown = 0;
 
-/**
```
Why did you delete all the documentation?

alphaville commented on src/main.cpp:
```diff
@@ -200,26 +165,26 @@ void loop()
 
-  // Compute control actions and send them to the motors
-
   controller.motorPwmSignals(attitudeError,
```
This function needs to be updated

alphaville commented on src/motors.cpp:
```diff
@@ -1,58 +1,89 @@
-    void MotorDriver::writeSpeedToEsc(int rotorSpeedFrontLeft, int rotorSpeedFrontRight,
-                                      int rotorSpeedBackLeft, int rotorSpeedBackRight)
-    {
-        m_frontLeftEsc.writeMicroseconds(rotorSpeedFrontLeft);   // sets the ESC speed
```
Put the documentation back

alphaville reviewed (changes requested):
@conquinn99 please see my comments. This is still incomplete. You need to update the control function (`controller.cpp` and `controller.hpp`). You'll also need to test this before we can merge it.

conquinn99 committed:
readded documentation on main.cpp and updated 'controller.motorPwmSignals' function to be suitable for a quadcopter or a hexacopter

conquinn99 committed:
readded documentation to motors.cpp

conquinn99 committed:
adjusted controller. cpp to include the two additional motors for a hexacopter (middle left and middle right) and adjusted the .hpp file to make this compilable

alphaville commented on src/controller.cpp:
```diff
@@ -54,11 +56,15 @@ namespace bzzz
         int mFR = throttle + controlToPwmScaling * (-controls[0] + controls[1] - controls[2]);
         int mBL = throttle + controlToPwmScaling * (controls[0] - controls[1] - controls[2]);
         int mBR = throttle + controlToPwmScaling * (-controls[0] - controls[1] + controls[2]);
+        int mML = throttle + controlToPwmScaling * (-controls[0]              - controls[2]); // adding for new motor middle left
```
The implementation needs to be completely different for quadcopters and hexacopters.

conquinn99 committed:
updated both controller.hpp and controller .cpp so that they should now be compatible for both quadcopter and hexacopter without breaking the quad for one or the other.

alphaville commented on src/controller.cpp:
```diff
@@ -42,6 +42,10 @@ namespace bzzz
         int &motorFR,
         int &motorBL,
         int &motorBR,
+        #if UAV_TYPE == UAV_TYPE_HEXACOPTER
```
Remove the indentation in front of all preprocessor directives

alphaville commented on include/controller.hpp:
```diff
@@ -86,6 +88,11 @@ the RC.
             int &motorBL,
             int &motorBR,
+            // when using hexacopter...
+        #if UAV_TYPE == UAV_TYPE_HEXACOPTER
```
Remove the indentation

alphaville commented on src/controller.cpp:
```diff
@@ -59,6 +63,15 @@ namespace bzzz
         motorFR = clip(mFR, motorClipLow, motorClipHigh);
         motorBL = clip(mBL, motorClipLow, motorClipHigh);
         motorBR = clip(mBR, motorClipLow, motorClipHigh);
+
```
This is not correct - check these notes: [Hexacopter v2.pdf](https://github.com/user-attachments/files/19665908/Hexacopter.v2.pdf)

alphaville commented on src/main.cpp:
```diff
@@ -22,17 +22,23 @@ float yawReferenceRad = 0.0;
  * be first acquiring its lock (i.e., the `timerMux`).
- */
+*/
+
```
Remove the blank line after the documentation

conquinn99 committed:
removed blank lines after documentation

conquinn99 committed:
removed indentation infront of preprocessor directives and updated the control action for quad and hexa use.

alphaville commented on src/controller.cpp:
```diff
@@ -42,23 +42,48 @@ namespace bzzz
         int &motorFR,
         int &motorBL,
         int &motorBR,
+        int &motorML,   // adding for new motor middle left
```
Don't you need a `#if UAV_TYPE == UAV_TYPE_HEXACOPTER` here?

conquinn99 committed:
corrected idle rotors when in quad-use, and removed #if preprocessor directive in function parameters

conquinn99 committed:
ensuring it compiles for both quad and hexa

alphaville commented on hardware/tests/hexa_arm_motors.cpp:
What is this test for? Add some comments at the beginning to explain what this test does.

alphaville commented on hardware/tests/test_single_motor.cpp:
Rename the test files: they should start with `hw_test_`

alphaville commented on include/config.hpp:
```diff
@@ -39,10 +44,19 @@
-#define FRONT_RIGHT_ESC_PIN 33
-#define BACK_LEFT_ESC_PIN 27
-#define BACK_RIGHT_ESC_PIN 26
+#if UAV_TYPE == UAV_TYPE_QUADCOPTER
```
I believe you can shorten this as follows:

```c++
#define FRONT_LEFT_ESC_PIN 25
#define FRONT_RIGHT_ESC_PIN 33
#define BACK_LEFT_ESC_PIN 27
#define BACK_RIGHT_ESC_PIN 26


#if UAV_TYPE == UAV_TYPE_HEXACOPTER
#define MIDDLE_LEFT_ESC_PIN 23   // temporary pin placement
#define MIDDLE_RIGHT_ESC_PIN 22  // temporary pin placement
#endif
```

Also, remove the indentation in front of all `#define`s

alphaville commented on include/motors.hpp:
```diff
@@ -36,12 +41,24 @@ namespace bzzz
+         * @param rotor_speed_middle_left set motor speed for MIDDLE_LEFT_ESC_PIN
+         * @param rotor_speed_middle_right set motor speed for MIDDLE_RIGHT_ESC_PIN
          */
+        #if UAV_TYPE == UAV_TYPE_HEXACOPTER
```
Remove the indentation from all preprocessor directives

alphaville commented on include/motors.hpp:
```diff
@@ -18,6 +18,11 @@ namespace bzzz
         Servo m_backLeftEsc;   // create servo object to control back left ESC
         Servo m_backRightEsc;  // create servo object to control back right ESC
 
+        #if UAV_TYPE == UAV_TYPE_HEXACOPTER
```
Remove the indentation from all preprocessor directives

alphaville commented on src/main.cpp:
```diff
@@ -211,15 +214,26 @@ void loop()
+                             yawRateReference,
+                             throttleRef,
+                             motorFL, motorFR, motorBL, motorBR,
+                             motorML, motorMR);
```
`motorML` and `motorMR` are not (or, at least, should not be) defined for quadcopters

alphaville commented on include/controller.hpp:
```diff
@@ -86,6 +88,8 @@ the RC.
             int &motorFR,
             int &motorBL,
             int &motorBR,
+            int &motorML,
```
`motorML` and `motorMR` are not (or, at least, should not be) defined for quadcopters

alphaville commented on src/controller.cpp:
```diff
@@ -42,23 +42,48 @@ namespace bzzz
         int &motorFR,
         int &motorBL,
         int &motorBR,
+        int &motorML,  // adding for new motor middle left
```
`motorML` and `motorMR` are not (or, at least, should not be) defined for quadcopters

alphaville commented on src/main.cpp:
```diff
@@ -22,17 +22,18 @@ float yawReferenceRad = 0.0;
 int motorFL, motorFR, motorBL, motorBR;
-bool wasKill=0;
-bool isKill=0;
+int motorML = 0, motorMR = 0; // Dummy for quad, used for hex
```
These variables shouldn't be defined at all for quadcopters

alphaville reviewed (changes requested):
@conquinn99, we should test this and see if the motors respond properly (needless to say, test without propellers on).

I don't know if `main.py` runs on startup on the Raspberry Pi. If not, have a look at [this documentation](https://github.com/QUB-ASL/bzzz/blob/main/raspberry/README.md). You can ask @jamie-54 to help you with the tuning.

conquinn99 committed:
shortened the motor config and also removed indentation

conquinn99 committed:
updated to remove incorrect indentation and also redefine motorML and motorMR when using either quad or hexacopter

conquinn99 committed:
changing the ML and MR pin numbers to the correct pins

alphaville commented:
@conquinn99 how is this going?

conquinn99 committed:
correct pins

conquinn99 committed:
Merge branch 'fix/refactoring-main' into feature/hexacopter-motors

Commit:
raspberry: main: fix sensor import

Fix import from pressure_sensor. Class was renamed to BMP180Sensor in previous refactor.

alphaville commented:
@conquinn99 can you please have a look at my comments above?

conquinn99 committed:
Separate quadcopter and hexacopter motorPwmSignals using UAV_TYPE_HEXACOPTER preprocessor

conquinn99 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/hexacopter-motors' into feature/hexacopter-motors

conquinn99 committed:
controller: quad vs hexa

alphaville commented:
@jamie-54 this looks more or less OK, but perhaps more testing is needed before we merge. Can you have a quick look and make sure you're happy with everything?

conquinn99 committed:
add 'UAV_TYPE=' to build flags

conquinn99 committed:
deletion of test_single_motor.cpp as it is not needed

conquinn99 committed:
fixing #if and #elif functions

conquinn99 committed:
Removed #define UAV_TYPE UAV_TYPE_HEXACOPTER as it can now be defined in platformio.ini file
